### PR TITLE
cli: add --dry-run flag to migrate apply (close #3128)

### LINKED ISSUE #3128: add a dry-run flag to migrate apply command
`hasura migrate apply` command should have a --dry-run flag, which will just print out the name/versions of the migrations that are going to be applied.
Usage:

```
# prints out the names of migrations are going to be applied
hasura migrate apply --dry-run
```

### PR DESCRIPTION
**Pull request by ashishra0**
### Description
Feature #3128 
Flag --dry-run is added to hasura migrate apply command in the CLI. Running the command with this flag will print out the migrations that will be applied. 

### Usage
```
$ hasura migrate apply --dry-run
INFO Migrations to be applied:
VERSION        Type  Name
1575986792321  up    create_table_public_testing

### CODE REVIEW

**Comment by shahidhk** [comment]
@ashishra0 Can you comment with a screenshot/paste of a sample output?

**Comment by arvi3411301** [comment]
@ashishra0 @shahidhk can we return an error if both `skip-execution` and `dry-run` flag are set to true?

File: cli/migrate/migrate.go [hunk 1]
```diff
@@ -1446,3 +1465,52 @@ func (m *Migrate) unlockErr(prevErr error) error {
...
+
+		switch r.(type) {
+		case error:
+			return r.(error)
+		case *Migration:
+			migr := r.(*Migration)
+			if migr.Body != nil {
+				version := int64(migr.Version)
+				fileName := migr.Identifier
+				fileType := convertBool(s.HasSuffix(fileName, "down.yaml"))
```
Comment (arvi3411301) [inline cli/migrate/migrate.go]: fileName can contain down.sql too. 

```golang
if migr.Version == migr.TargetVersion {
    return "up"
} else {
    return "down"
}
```

File: cli/commands/migrate_apply.go [hunk 2]
```diff
@@ -25,13 +25,20 @@ func newMigrateApplyCmd(ec *cli.ExecutionContext) *cobra.Command {
 			return ec.Validate()
 		},
 		RunE: func(cmd *cobra.Command, args []string) error {
-			opts.EC.Spin("Applying migrations...")
+			if opts.dryRun && opts.skipExecution {
+				return errors.New("both --skip-execution and --dry-run flag cannot be set to true")
```
Comment (shahidhk) [inline cli/commands/migrate_apply.go]:
```suggestion
				return errors.New("both --skip-execution and --dry-run flag cannot be used together")
```

File: cli/migrate/migrate.go [hunk 3]
```diff
@@ -1101,6 +1119,32 @@ func (m *Migrate) runMigrations(ret <-chan interface{}) error {
...
+			if migr.Body != nil {
+				version := int64(migr.Version)
+				if version != lastInsertVersion {
+					migrations = append(migrations, migr)
+					lastInsertVersion = version
+				}
+			}
+		}
+	}
+	fmt.Println(printDryRunStatus(migrations).String())
```
Comment (shahidhk) [inline cli/migrate/migrate.go]: use the logger instead of fmt.

**Comment by shahidhk** [comment]
@ashishra0 Can you test it out once again and post an output?

**Comment by ashishra0** [comment]
@shahidhk The logger is working as expected. Here is the output it gives

```
INFO VERSION        TYPE  NAME
1576752677865  up    alter_table_public_recipes_add_column_title
```

**Comment by danazkari** [comment]
What is this PR missing? If there's something I can help with @ashishra0 please let me know :) I'm really interested in this feature for my projects (specially for creating a github-action that comments the results of the dry run to a PR)

**Pull request merged**